api/controllers/groupAdmin.js

```diff
@@ -0,0 +1,314 @@
+                    as: "messages"
+                }
+            },
+            {
+                $lookup: {
+                    from: "members",
+                    localField: "group.id",
+                    foreignField: "groupId",
+                    as: "members"
+                }
+            },
+            {
+                $project: {
+                    members: {$size: "$members"},
+                    messages: {$size: "$messages"}
+                }
+            }
+        ], (err, result) => {
+            if (err) res.status(500).send(err)
+            members = result[0].members;
+            messages = result[0].messages;
+            mongoCollections.mongoMessages.find({'message.chat.id': group_id}).toArray((err, result)=> {
+                if (err) res.status(500).send(err)
+                res.send({"actions": result.filter(x => x.message.entities != null).length, "members": members, "messages": messages})
+            })
+        });
+        
+    }
+}
+
+exports.getGroupMessages = (mongoCollections) => {
+    return (req, res) => {
+        group_id = parseInt(req.body.group_id)
+        mongoCollections.mongoMessages.aggregate([
+            {
+                $match: {
+                    'message.chat.id': group_id
+                }
+            },
+            {
+                $sort: {
+                    'postedDate': -1
+                }
+            }
+        ], (err, result)=> {
+            if (err) res.status(500).send(err)
+            res.send(result)
+        });
+    }
+}
+
+exports.getGroupMembers = (mongoCollections) => {
+    return (req, res) => {
+        group_id = parseInt(req.body.group_id);
+        mongoCollections.mongoGroupMembers.find({groupId: group_id}).toArray((err, result) => {
+            if (err) res.status(500).send(err)
+            res.send(result)
+        });
+    }
+}
+
+exports.getGroupMessagesYearlyCount = (mongoCollections) => {
+    return (req, res) => {
+        group_id = parseInt(req.body.group_id);
+        mongoCollections.mongoMessages.aggregate([
+            {
+                $match: {
+                    'message.chat.id': group_id
+                }
+            },
+            {
+                $project: {
+                    month: {"$month": '$postedDate'},
+                    messageYear: {"$year": "$postedDate"}
+                }
+            },
+            {
+                $group: {
+                    _id: {month: "$month", year: "$messageYear"},
+                    messages: {$sum: 1}
+                }
+            }
+        ], (err, result) => {
+            if (err) res.status(500).send(err);
+                res.send(result);
+        })
+    }
+}
+
+exports.getGroupMessagesMonthlyCount = (mongoCollections) => {
+    return (req, res) => {
+        group_id = parseInt(req.body.group_id);
+        var month = (new Date()).getMonth() + 1;
+        mongoCollections.mongoMessages.aggregate([
```

Comment:
Actually it will be much better if you will write manipulations with mongoose in async\await way.

Reply:
That'll have to open a new connection to database which I don't think is a good idea when there's already an opened connection to be used. Please correct me if I'm wrong.

Reply:
I did not understand how rewriting this code on async can have an effect on the way how it works.

Do you familiar with async/await?
Async/await way of writing asynchronous code is the equivalent to callbacks/promises way but just looks different syntactically. Mongoose allow to write code in promises way (which is actually async way). So I just want you to write it on async.

Reply:
Alright let me change the code and come back for review.